internal/idp/providers/github/github.go

```diff
@@ -0,0 +1,177 @@
+package github
+
+import (
+	"strconv"
+	"time"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/text/language"
+
+	"github.com/zitadel/zitadel/internal/idp"
+	"github.com/zitadel/zitadel/internal/idp/providers/oauth"
+)
+
+const (
+	authURL    = "https://github.com/login/oauth/authorize"
+	tokenURL   = "https://github.com/login/oauth/access_token"
+	profileURL = "https://api.github.com/user"
+	name       = "GitHub"
+)
+
+var _ idp.Provider = (*Provider)(nil)
+
+// New creates a GitHub.com provider using the oauth.Provider (OAuth 2.0 generic provider)
+func New(clientID, secret, callbackURL string, scopes []string, options ...oauth.ProviderOpts) (*Provider, error) {
+	return NewCustomURL(name, clientID, secret, callbackURL, authURL, tokenURL, profileURL, scopes, options...)
+}
+
+// NewCustomURL creates a GitHub provider using the oauth.Provider (OAuth 2.0 generic provider)
+// with custom endpoints, e.g. GitHub Enterprise server
+func NewCustomURL(name, clientID, secret, callbackURL, authURL, tokenURL, profileURL string, scopes []string, options ...oauth.ProviderOpts) (*Provider, error) {
+	rp, err := oauth.New(
+		newConfig(clientID, secret, callbackURL, authURL, tokenURL, scopes),
+		name,
+		profileURL,
+		func() oauth.UserInfoMapper {
+			return new(User)
+		},
+		options...,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return &Provider{
+		Provider: rp,
+	}, nil
+}
+
+// Provider is the idp.Provider implementation for GitHub
+type Provider struct {
+	*oauth.Provider
+}
+
+func newConfig(clientID, secret, callbackURL, authURL, tokenURL string, scopes []string) *oauth2.Config {
+	c := &oauth2.Config{
+		ClientID:     clientID,
+		ClientSecret: secret,
+		RedirectURL:  callbackURL,
+		Endpoint: oauth2.Endpoint{
+			AuthURL:  authURL,
+			TokenURL: tokenURL,
+		},
+		Scopes: scopes,
+	}
+
+	return c
+}
+
+// User is a representation of the authenticated GH user
+// https://docs.github.com/en/rest/users/users?apiVersion=2022-11-28#get-the-authenticated-user
```

Comment:
go recommends to treat links as references. check out the links section here: https://tip.golang.org/doc/comment